src/librustc_mir/monomorphize/polymorphize.rs

```diff
@@ -0,0 +1,283 @@
+//! Polymorphization Analysis
+//! =========================
+//!
+//! This module implements an analysis of functions, methods and closures to determine which
+//! generic parameters are unused (and eventually, in what ways generic parameters are used - only
+//! for their size, offset of a field, etc.).
+
+use rustc::mir::{
+    visit::{TyContext, Visitor},
+    CastKind, Local, LocalDecl, Location, ReadOnlyBodyAndCache, Rvalue, TerminatorKind,
+};
+use rustc::ty::{
+    self,
+    adjustment::PointerCast,
+    fold::{TypeFoldable, TypeVisitor},
+    query::Providers,
+    Ty, TyCtxt,
+};
+use rustc_hir::{self as hir, def_id::DefId};
+use rustc_index::bit_set::BitSet;
+use rustc_span::symbol::Symbol;
+use rustc_target::spec::abi::Abi;
+
+/// Provide implementations of queries relating to polymorphization analysis.
+pub fn provide(providers: &mut Providers<'_>) {
+    providers.used_generic_parameters = used_generic_parameters;
+}
+
+/// Determine which generic parameters are used by the function/method/closure represented by
+/// `def_id`. Returns a `BitSet` with a domain size matching the number of generic parameters,
+/// set bits represent that the matching generic parameter is used.
+fn used_generic_parameters(tcx: TyCtxt<'_>, def_id: DefId) -> BitSet<u32> {
+    debug!("used_generic_parameters({:?})", def_id);
+    let generics = tcx.generics_of(def_id);
+    debug!("used_generic_parameters: generics={:?}", generics);
+
+    if !tcx.sess.opts.debugging_opts.polymorphize {
+        // If the analysis is not enabled, then return a `BitSet` saying that every generic
+        // parameter was used.
+        return BitSet::new_filled(generics.count());
+    }
+
+    // If the item is not local then say every generic parameter was used.
```

Comment:
You probably want to make this work cross-crate, sooner than later (since it will likely expose more bugs more quickly).

The simplest thing IMO would be to check if the MIR is available, but long-term we'll probably want to save the result of this analysis and not recompute it.

Reply:
Fixed in 0d7944d.